src/RealPath.php

```diff
@@ -24,74 +23,28 @@
 
 /**
  * @psalm-type Options = array{
- *     exists: bool,
- *     ...
+ *     exists?: bool,
  * }
- * @template TOptions of Options
- * @extends AbstractFilter<TOptions>
+ * @implements FilterInterface<string|array<array-key, string|mixed>>
  */
-final class RealPath extends AbstractFilter
+final class RealPath implements FilterInterface
 {
-    /** @var TOptions $options */
-    protected $options = [
-        'exists' => true,
-    ];
+    private readonly bool $pathMustExist;
 
-    /**
-     * @param  bool|Traversable|Options $existsOrOptions Options to set
-     */
-    public function __construct($existsOrOptions = true)
+    /** @param Options $options */
+    public function __construct(array $options = [])
     {
-        if ($existsOrOptions !== null) {
-            if (! static::isOptions($existsOrOptions)) {
-                $this->setExists($existsOrOptions);
-            } else {
-                $this->setOptions($existsOrOptions);
-            }
-        }
-    }
-
-    /**
-     * Sets if the path has to exist
-     * TRUE when the path must exist
-     * FALSE when not existing paths can be given
-     *
-     * @param  bool $flag Path must exist
-     * @return self
-     */
-    public function setExists($flag = true)
-    {
-        $this->options['exists'] = (bool) $flag;
-        return $this;
+        $this->pathMustExist = $options['exists'] ?? true;
     }
 
-    /**
-     * Returns true if the filtered path must exist
-     *
-     * @return bool
-     */
-    public function getExists()
-    {
-        return $this->options['exists'];
-    }
-
-    /**
-     * Defined by Laminas\Filter\FilterInterface
-     *
-     * Returns realpath($value)
-     *
-     * If the value provided is non-scalar, the value will remain unfiltered
-     *
-     * @psalm-return ($value is string ? string : mixed)
-     */
     public function filter(mixed $value): mixed
     {
         if (! is_string($value)) {
             return $value;
         }
         $path = (string) $value;
 
-        if ($this->options['exists']) {
+        if ($this->pathMustExist) {
             return realpath($path);
```

Comment:
Typical filter behaviour should be to either return the successfully filtered value or the un-filtered value. This is potentially returning `false`, so I think there should be conditional to return `$value` when `realpath() === false`.

```php
// file_exists will return appropriate result for args such as '/some/../other/../foo/bar'
if ($this->pathMustExist && ! file_exists($value)) {
    return $value;
}

if ($this->pathMustExist) {
    return realpath($value);
}
```

This behaviour change will need a sentence in the migration docs and the tests will need changing too.

Reply:
Yes! Improving docs to state that filters shouldn't be used for validation is a worthy improvement! As filter does not depend on validator, I don't think its _necessary_ to mention laminas-validator itself, but write whatever makes most sense to the user - Typically filtering happens prior to validation, so returning the un-modified path argument when it is un-filterable is good, because it should be validated by `..\File\Exists` afterwards.

Reply:
Cool. Thanks